rom 133afb15e0c4b0fd1c6e194b8193ff369042c321 Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Wed, 14 Apr 2021 15:13:08 +0200
Subject: [PATCH] bootscript-yocto.txt: Allow flashing, integrate balenaOS

Integrate boot script with BalenaOS and allow flashing
the eMMC.

Upstream-Status: Inappropriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 .../boundary/bootscripts/bootscript-yocto.txt | 59 ++++++++++++++++---
 1 file changed, 51 insertions(+), 8 deletions(-)

diff --git a/board/boundary/bootscripts/bootscript-yocto.txt b/board/boundary/bootscripts/bootscript-yocto.txt
index 92f3cbae46..63d49fb479 100644
--- a/board/boundary/bootscripts/bootscript-yocto.txt
+++ b/board/boundary/bootscripts/bootscript-yocto.txt
@@ -26,7 +26,7 @@ fi
 
 setexpr a_script  ${a_base} + 0x00800000
 setexpr a_zImage  ${a_base} + 0x00800000
-setexpr a_fdt     ${a_base} + 0x03000000
+setexpr a_fdt     ${a_base} + 0x03300000
 setexpr a_initrd  ${a_base} + 0x03100000
 setexpr a_reset_cause_marker ${a_base} + 0x80
 setexpr a_reset_cause	     ${a_base} + 0x84
@@ -72,6 +72,19 @@ if itest.s x${distro_bootpart} == x ; then
 	distro_bootpart=1
 fi
 
+if mmc dev 1; then
+    setenv devnum 1;
+    echo Using SD Card to load BalenaOS;
+fi;
+
+setenv resin_dev_index ${devnum};
+setenv resin_kernel_load_addr ${a_zImage}
+
+if env exists resin_set_kernel_root; then
+    run resin_set_kernel_root;
+    run set_os_cmdline;
+fi;
+
 if load ${devtype} ${devnum}:${distro_bootpart} ${a_script} uEnv.txt ; then
     env import -t ${a_script} ${filesize}
 fi
@@ -87,12 +100,24 @@ if itest.s x${consoleblank} == x ; then
 fi
 setenv bootargs ${bootargs} vmalloc=${vmalloc} consoleblank=${consoleblank} rootwait fixrtc cpu=${imx_cpu} board=${board} uboot_release=${uboot_release}
 
-if load ${devtype} ${devnum}:${distro_bootpart} ${a_fdt} ${prefix}${fdt_file} ; then
-	fdt addr ${a_fdt}
+if env exists resin_set_kernel_root; then
+        if load ${devtype} ${devnum}:${resin_root_part} ${a_fdt} boot/${fdt_file} ; then
+                fdt addr ${a_fdt}
+        elif load ${devtype} ${devnum}:${resin_boot_part} ${a_fdt} ${fdt_file} ; then
+                echo "Did not find dtb in root filesystem, loaded it from boot partition"
+                fdt addr ${a_fdt}
+        else
+                echo "!!!! Error loading Balena ${prefix}${fdt_file}";
+                exit;
+        fi
 else
-	echo "!!!! Error loading ${prefix}${fdt_file}";
-	exit;
-fi
+        if load ${devtype} ${devnum}:${distro_bootpart} ${a_fdt} ${prefix}${fdt_file} ; then
+                fdt addr ${a_fdt}
+        else
+                echo "!!!! Error loading ${prefix}${fdt_file}";
+                exit;
+        fi;
+fi;
 
 fdt resize 4096
 if itest.s "x" != "x${cmd_board}" ; then
@@ -132,7 +157,11 @@ if test "sata" = "${devtype}" ; then
 elif test "usb" = "${devtype}" ; then
 	setenv bootargs "${bootargs} root=/dev/sda${bpart}" ;
 else
-	setenv bootargs "${bootargs} root=/dev/mmcblk${devnum}p${bpart}"
+        if env exists resin_set_kernel_root; then
+                setenv bootargs "${bootargs} ${resin_kernel_root} ro rootwait ";
+        else
+                setenv bootargs "${bootargs} root=/dev/mmcblk${devnum}p2 ro rootwait flasher";
+        fi;
 fi
 
 if itest.s "x" != "x${disable_msi}" ; then
@@ -177,7 +206,21 @@ if itest.s "x" != "x${show_env}" ; then
 	printenv
 fi
 
-if load ${devtype} ${devnum}:${distro_bootpart} ${a_zImage} ${prefix}${kernelimage} ; then
+if env exists resin_set_kernel_root; then
+    setenv bootargs "${bootargs} ${os_cmdline} "
+fi;
+
+if env exists resin_set_kernel_root; then
+    setenv bootargs "${bootargs} ${os_cmdline} "
+
+    if load ${devtype} ${resin_dev_index}:${resin_root_part} ${a_zImage} boot/${kernelimage} ; then
+        ${bootcommand} ${a_zImage} - ${a_fdt}
+    fi;
+
+    echo "Error loading Balena kernel image"
+fi;
+
+if load ${devtype} ${devnum}:2 ${a_zImage} boot/${kernelimage} ; then
 	${bootcommand} ${a_zImage} - ${a_fdt}
 fi
 echo "Error loading kernel image"
-- 
2.37.2

